File: 프로그래머스/lv3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    
    int answer = 0;
    int dp[101][101] = {1, };
    

    //초기화
    for(int i = 0; i<101; i++)
        for(int j = 0; j<101; j++)
            dp[i][j] = -1;
    
    //웅덩이 설정
    for(int i = 0; i<puddles.size(); i++)
    {
        dp[puddles[i][1]][puddles[i][0]] = 0;
    }

    //탐색
    dp[1][1] = 1;
    for(int i = 1; i<= n; i++)
    {
        for(int j = 1; j<=m; j++)
        {
            if(dp[i][j] == -1)
            {
                int temp1, temp2;
                if(i > 1)
                    temp1 = dp[i-1][j];
                else
                    temp1 = 0;

                if( j > 1)
                    temp2 = dp[i][j-1];
                else
                    temp2 = 0;
                
                dp[i][j] = (temp1 + temp2) % 1000000007;
            }
        }
    }

    answer = dp[n][m];
    
    return answer;
}

File: 프로그래머스/lv3/42898. 등굣길/등굣길.java

class Solution {

	public int solution(int m, int n, int[][] puddles) {
        int[][] dp = new int[n+1][m+1];
        int[][] map = new int[n+1][m+1];
        
        for(int i = 0; i < puddles.length; i++)
        	map[puddles[i][1]][puddles[i][0]] = -1;
        
        dp[1][1] = 1;
        
        for(int i = 1; i <= n; i++)
        	for(int j = 1; j <= m; j++) {
     
        		if(map[i][j-1] != -1)
        			dp[i][j] += dp[i][j-1];
        		
        		if(map[i-1][j] != -1)
        			dp[i][j] += dp[i-1][j];
        		
        		dp[i][j] = dp[i][j] % 1000000007;
        	}
        
        return dp[n][m];
    }
    
    
    public static void main(String args[]) {
    	
    	System.out.println(new Solution().solution(4,3,new int[][] {{2,2}}));
    }
}